File: desktop-viewer/platform-adb/adb-core/src/main/kotlin/com/androidperformancestudio/platform/adb/AdbClient.kt
package com.androidperformancestudio.platform.adb

import java.nio.file.Files
import java.nio.file.Path
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

interface AdbClient {
    suspend fun listDevices(): List<AdbDevice>

    suspend fun shell(
        serial: String,
        arguments: List<String>,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbTextResult

    suspend fun execOut(
        serial: String,
        arguments: List<String>,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbBinaryResult

    suspend fun push(
        serial: String,
        localPath: Path,
        remotePath: String,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbTextResult

    suspend fun pull(
        serial: String,
        remotePath: String,
        localPath: Path,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbTextResult

    suspend fun forward(
        serial: String,
        local: String,
        remote: String,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbTextResult

    suspend fun removeForward(
        serial: String,
        local: String,
        timeout: Duration = DEFAULT_TIMEOUT,
    ): AdbTextResult

    companion object {
        val DEFAULT_TIMEOUT: Duration = 30.seconds
    }
}

class DefaultAdbClient(
    private val executable: Path,
    private val processRunner: ProcessRunner = JvmProcessRunner(),
    private val devicesParser: AdbDevicesParser = AdbDevicesParser(),
) : AdbClient {
    override suspend fun listDevices(): List<AdbDevice> {
        val command = command(listOf("devices", "-l"), AdbClient.DEFAULT_TIMEOUT)
        return devicesParser.parse(processRunner.executeText(command).requireSuccess(command).stdout)
    }

    override suspend fun shell(
        serial: String,
        arguments: List<String>,
        timeout: Duration,
    ): AdbTextResult = executeText(adbDeviceArguments(serial, "shell", arguments), timeout)

    override suspend fun execOut(
        serial: String,
        arguments: List<String>,
        timeout: Duration,
    ): AdbBinaryResult = executeBinary(adbDeviceArguments(serial, "exec-out", arguments), timeout)

    override suspend fun push(
        serial: String,
        localPath: Path,
        remotePath: String,
        timeout: Duration,
    ): AdbTextResult {
        if (!Files.isRegularFile(localPath)) {
            throw AdbInputException("Local push source is not a file: $localPath")
        }
        return executeText(
            adbDeviceArguments(
                serial,
                "push",
                listOf(localPath.toString(), AdbInputValidator.requireRemotePath(remotePath)),
            ),
            timeout,
        )
    }

    override suspend fun pull(
        serial: String,
        remotePath: String,
        localPath: Path,
        timeout: Duration,
    ): AdbTextResult =
        executeText(
            adbDeviceArguments(
                serial,
                "pull",
                listOf(AdbInputValidator.requireRemotePath(remotePath), localPath.toString()),
            ),
            timeout,
        )

    override suspend fun forward(
        serial: String,
        local: String,
        remote: String,
        timeout: Duration,
    ): AdbTextResult =
        executeText(
            adbDeviceArguments(
                serial,
                "forward",
                listOf(
                    AdbInputValidator.requireForwardEndpoint(local),
                    AdbInputValidator.requireForwardEndpoint(remote),
                ),
            ),
            timeout,
        )

    override suspend fun removeForward(
        serial: String,
        local: String,
        timeout: Duration,
    ): AdbTextResult =
        executeText(
            adbDeviceArguments(
                serial,
                "forward",
                listOf("--remove", AdbInputValidator.requireForwardEndpoint(local)),
            ),
            timeout,
        )

    private suspend fun executeText(
        arguments: List<String>,
        timeout: Duration,
    ): AdbTextResult {
        val command = command(arguments, timeout)
        return processRunner.executeText(command).requireSuccess(command)
    }

    private suspend fun executeBinary(
        arguments: List<String>,
        timeout: Duration,
    ): AdbBinaryResult {
        val command = command(arguments, timeout)
        return processRunner.executeBinary(command).requireSuccess(command)
    }

    private fun command(
        arguments: List<String>,
        timeout: Duration,
    ): AdbCommand = AdbCommand(executable = executable, arguments = arguments, timeout = timeout)
}

private fun adbDeviceArguments(
    serial: String,
    operation: String,
    arguments: List<String>,
): List<String> = listOf("-s", AdbInputValidator.requireSerial(serial), operation) + arguments

File: electron-viewer/packages/platform-adb/src/adb-client.ts
import { statSync } from 'node:fs';
import {
  HostProcessCancelledError,
  HostProcessStartError,
  HostProcessTimeoutError,
  runHostProcessBinary,
  type HostProcessBinaryResult,
  type HostProcessRequest,
} from '@aps/platform-host';
import { DEFAULT_ADB_MAX_OUTPUT_BYTES, DEFAULT_ADB_TIMEOUT_MS, buildAdbDeviceArgs, quoteRemoteShellArgument } from './adb-command.js';
import { parseAdbDevices, type AdbDevice } from './adb-device.js';
import {
  AdbCommandCancelledError,
  AdbCommandFailedError,
  AdbCommandTimeoutError,
  AdbInputError,
  AdbProcessStartError,
} from './adb-errors.js';
import { requireForwardEndpoint, requireRemotePath } from './adb-input-validator.js';

export type AdbBinaryExecutor = (request: HostProcessRequest) => Promise<HostProcessBinaryResult>;

export interface AdbTextResult {
  readonly pid: number;
  readonly exitCode: number;
  readonly stdout: string;
  readonly stderr: string;
  readonly durationMs: number;
  readonly stdoutTruncated: boolean;
  readonly stderrTruncated: boolean;
}

export interface AdbOptions {
  readonly timeoutMs?: number;
  readonly maxOutputBytesPerStream?: number;
  readonly signal?: AbortSignal;
}

export interface AdbClientOptions {
  readonly executable: string;
  readonly execute?: AdbBinaryExecutor;
  readonly timeoutMs?: number;
  readonly maxOutputBytesPerStream?: number;
  readonly isRegularFile?: (path: string) => boolean;
}

function defaultIsRegularFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

export class AdbClient {
  private readonly executable: string;
  private readonly execute: AdbBinaryExecutor;
  private readonly timeoutMs: number;
  private readonly maxOutputBytesPerStream: number;
  private readonly isRegularFile: (path: string) => boolean;

  constructor(options: AdbClientOptions) {
    this.executable = options.executable;
    this.execute = options.execute ?? runHostProcessBinary;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_ADB_TIMEOUT_MS;
    this.maxOutputBytesPerStream = options.maxOutputBytesPerStream ?? DEFAULT_ADB_MAX_OUTPUT_BYTES;
    this.isRegularFile = options.isRegularFile ?? defaultIsRegularFile;
  }

  async listDevices(options: AdbOptions = {}): Promise<AdbDevice[]> {
    const result = await this.runText(['devices', '-l'], options);
    return parseAdbDevices(result.stdout);
  }

  async shell(serial: string, args: readonly string[], options: AdbOptions = {}): Promise<AdbTextResult> {
    return this.runText(
      buildAdbDeviceArgs(serial, 'shell', args.map((argument) => quoteRemoteShellArgument(argument))),
      options,
    );
  }

  async execOut(serial: string, args: readonly string[], options: AdbOptions = {}): Promise<HostProcessBinaryResult> {
    return this.runBinary(buildAdbDeviceArgs(serial, 'exec-out', args), options);
  }

  async push(serial: string, localPath: string, remotePath: string, options: AdbOptions = {}): Promise<AdbTextResult> {
    if (!this.isRegularFile(localPath)) {
      throw new AdbInputError('Local push source is not a file: ' + localPath);
    }
    return this.runText(buildAdbDeviceArgs(serial, 'push', [localPath, requireRemotePath(remotePath)]), options);
  }

  async pull(serial: string, remotePath: string, localPath: string, options: AdbOptions = {}): Promise<AdbTextResult> {
    return this.runText(buildAdbDeviceArgs(serial, 'pull', [requireRemotePath(remotePath), localPath]), options);
  }

  async forward(serial: string, local: string, remote: string, options: AdbOptions = {}): Promise<AdbTextResult> {
    return this.runText(
      buildAdbDeviceArgs(serial, 'forward', [requireForwardEndpoint(local), requireForwardEndpoint(remote)]),
      options,
    );
  }

  async removeForward(serial: string, local: string, options: AdbOptions = {}): Promise<AdbTextResult> {
    return this.runText(buildAdbDeviceArgs(serial, 'forward', ['--remove', requireForwardEndpoint(local)]), options);
  }

  async bugreport(serial: string, outputPath: string, options: AdbOptions = {}): Promise<AdbTextResult> {
    return this.runText(buildAdbDeviceArgs(serial, 'bugreport', [outputPath]), options);
  }

  private request(args: readonly string[], options: AdbOptions): HostProcessRequest {
    return {
      executable: this.executable,
      args,
      timeoutMs: options.timeoutMs ?? this.timeoutMs,
      maxOutputBytesPerStream: options.maxOutputBytesPerStream ?? this.maxOutputBytesPerStream,
      ...(options.signal !== undefined ? { signal: options.signal } : {}),
    };
  }

  private async runBinary(args: readonly string[], options: AdbOptions): Promise<HostProcessBinaryResult> {
    const request = this.request(args, options);
    const command = [request.executable, ...args];
    let result: HostProcessBinaryResult;
    try {
      result = await this.execute(request);
    } catch (error) {
      if (error instanceof HostProcessStartError) throw new AdbProcessStartError(command, error);
      if (error instanceof HostProcessTimeoutError) throw new AdbCommandTimeoutError(command, error.timeoutMs, error.pid);
      if (error instanceof HostProcessCancelledError) throw new AdbCommandCancelledError(command, error.pid);
      throw error;
    }
    if (result.exitCode !== 0) {
      throw new AdbCommandFailedError(command, result.exitCode, result.stderr.toString('utf8'));
    }
    return result;
  }

  private async runText(args: readonly string[], options: AdbOptions): Promise<AdbTextResult> {
    const result = await this.runBinary(args, options);
    return {
      pid: result.pid,
      exitCode: result.exitCode,
      stdout: result.stdout.toString('utf8'),
      stderr: result.stderr.toString('utf8'),
      durationMs: result.durationMs,
      stdoutTruncated: result.stdoutTruncated,
      stderrTruncated: result.stderrTruncated,
    };
  }
}
